### PR TITLE
Hangdu issue5

### LINKED ISSUE #5: Image Tags
#### As a user I want to add tags to images.

__Story__:

Our users are loving the ability to add images to our service. However, now
that there are more than a handful of images, it is apparent that we need some
way to filter images by certain categories. In order to do that, we first need
to label images in some way. It is time to add the ability for our users to tag
images when they add them.

__Acceptance criteria__:
- [x] When I add a new image, there is a form field to add tags.
- [x] After I've saved an image with tags, I see those tags displayed with the
  associated image.
- [x] My application uses the acts-as-taggable-on gem to provide the tagging
  functionality.

__Reference material__:
- https://github.com/mbleigh/acts-as-taggable-on

__Note__: You should allow images to have no tags; this sets the stage for a later issue.

__Dependencies__:
- Image Index

### PR DESCRIPTION
**Pull request by HangDu1995**
Closes #5 

---

The following is a checklist to remind you of the steps to take when opening a
new PR. The horizontal line above and everything below can be removed from the
pull request description.

- [ ] I have replaced `X` above with the associated issue number.
- [ ] I have changed the issue's label to "pull request".
- [ ] I have assigned the designated reviewer to this PR.
- [ ] I have set the appropriate label on this PR.
- [ ] I have written clear commit messages ([ref](http://chris.beams.io/posts/git-commit/)).

If needed, please review the document on the
[Image Sharing Project](https://sites.google.com/a/appfolio.com/eng/new-hire-resources/engineering-academy/ropes/image-sharing-project)
for information on how labels should be used.

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -6,6 +6,13 @@ AllCops:
     - 'vendor/**/*'
     - 'node_modules/**/*'
     - 'bin/*'
+    - 'db/migrate/20190910050915_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb'
```
Comment (choruk) [inline .rubocop.yml]: I'm pretty sure you can write this as `db/migrate/*.acts_as_taggable_on_engine.rb` instead of listing each migration individually.

File: test/controllers/images_controller_test.rb [hunk 2]
```diff
@@ -48,31 +48,68 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+      assert_select '.tag' do |elements|
+        assert_equal 2, elements.length
+        assert_equal 'dog', elements[0].text
+        assert_equal 'cute', elements[1].text
+      end
+    end
 
-    assert_select 'p' do |tags|
-      assert_equal 'dog, cute', tags[0]
+    # tags also show on 'index' page for all images
```
Comment (choruk) [inline test/controllers/images_controller_test.rb]: While this certainly works, it is generally better to split the testing of different controller actions into separate test cases. That way if something is broken on both the show and index page, you will get two separate test failures and might see a common pattern instead of only getting the first failure and then not seeing the second one until you fix that.

File: test/controllers/images_controller_test.rb [hunk 3]
```diff
@@ -48,31 +48,68 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+    assert_select('.tags').first do
+      assert_select '.tag' do |elements|
+        assert_equal 2, elements.length
+        assert_equal 'dog', elements[0].text
+        assert_equal 'cute', elements[1].text
+      end
     end
+  end
+
+  test 'should display None when no tags' do
```
Comment (choruk) [inline test/controllers/images_controller_test.rb]: If this test is about verifying that `None` is displayed when there is an image with no tags, there is no need to do a `post` to create the image. You can just do `Image.create!(...)` to create the image. If you want to test that you can make a `post` without any tags and still create an image, that should be split out into a separate test case.

File: app/views/images/index.html.erb [hunk 4]
```diff
@@ -1,11 +1,34 @@
 <h1>Listing Images</h1>
-<%= link_to 'New Image', new_image_path %>
 
+<p>
+  <%= link_to 'New Image', new_image_path %>
+</p>
+
+<p>
+  <%= '----------------------------------------------------------' %>
```
[the remark this answers is not in the document]
Comment (HangDu1995) [reply app/views/images/index.html.erb]: I added another test case for the situation where the link field is blank and asserted if tags are kept. Since my only two inputs are link and tags I think that plus other tests should cover different user inputs. And I changed <%= '---------' %> to <p>
  ---------
</p>

File: .rubocop.yml (hunk 1 above)
Comment (HangDu1995) [reply .rubocop.yml]: Yes you are right. I fixed that

File: test/controllers/images_controller_test.rb (hunk 2 above)
Comment (HangDu1995) [reply test/controllers/images_controller_test.rb]: I split test cases for index and show action

File: test/controllers/images_controller_test.rb (hunk 3 above)
Comment (HangDu1995) [reply test/controllers/images_controller_test.rb]: Added test 'should create an image without tags' and split this test into tests on show and index page separately

File: test/controllers/images_controller_test.rb [hunk 5]
```diff
@@ -48,31 +49,106 @@ class ImagesControllerTest < ActionDispatch::IntegrationTest
...
+      end
     end
+  end
 
+  test 'should create an image without tags' do
+    image_link = 'https://petlandstl.com/wp-content/themes/cosmick-petland-global/images/cta1-1.jpg'
+    no_tag = nil
+
+    assert_difference 'Image.count', 1 do
+      assert Image.create!(link: image_link, tag_list: no_tag)
```
Comment (choruk) [inline test/controllers/images_controller_test.rb]: Since you aren't actually making any requests to the controller in this test, it shouldn't be in the controller test file. I think you already have a model test covering this case though, so you could either delete this test or change it to use `post` with an empty tag list.

Comment (HangDu1995) [reply test/controllers/images_controller_test.rb]: Ok, deleted it
